### PR TITLE
Add metadata_log_entries table to Iceberg

### PR DESCRIPTION
**Pull request by oneonestar**
## Description
Add metadata_log_entries table to Iceberg

```
SELECT * FROM "test_table$metadata_log_entries"

             timestamp                 |                                                              file                                                          | latest_snapshot_id  | latest_schema_id | latest_sequence_number
---------------------------------------+----------------------------------------------------------------------------------------------------------------------------+---------------------+------------------+------------------------
 2024-01-16 15:55:31.172 Europe/Vienna | hdfs://hadoop-master:9000/user/hive/warehouse/test_table/metadata/00000-39174715-be2a-48fa-9949-35413b8b736e.metadata.json | 1221802298419195590 |                0 |                      1
 2024-01-16 17:19:56.118 Europe/Vienna | hdfs://hadoop-master:9000/user/hive/warehouse/test_table/metadata/00001-e40178c9-271f-4a96-ad29-eed5e7aef9b0.metadata.json | 7124386610209126943 |                0 |                      2
```

## Additional context and related issues
Reference:
https://iceberg.apache.org/docs/latest/spark-queries/#metadata-log-entries
https://www.dremio.com/blog/getting-started-with-flink-sql-and-apache-iceberg/#h-metadata-log-entries

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Add support for `metadata_log_entries` system table. ({issue}`20410`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        return tableMetadata;
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        return new FixedPageSource(buildPages(tableMetadata, session, icebergTable));
+    }
+
+    private static List<Page> buildPages(ConnectorTableMetadata tableMetadata, ConnectorSession session, Table icebergTable)
```
[the remark this answers is not in the document]
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:94]: Same issue in Spark 3.3.x and Trino.
I don't know is this a correct behavior.

```
spark-sql> SELECT * FROM test.test_table.metadata_log_entries;
2024-01-22 16:50:33.208	hdfs://hadoop-master:9000/metadata/00000-acbc6857-46cd-498a-b8e0-95b5a8b661df.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:39.595	hdfs://hadoop-master:9000/metadata/00001-9a9310e3-b54c-4bb1-a9ae-f8a775b51755.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:40.7	hdfs://hadoop-master:9000/metadata/00002-72310e31-4d95-4f60-ba43-9f4e336f1ff9.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:43.645	hdfs://hadoop-master:9000/metadata/00003-ef40bea2-3719-4174-890b-d3aced094ba4.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:44.734	hdfs://hadoop-master:9000/metadata/00004-9698fb96-0ea7-48b4-a499-4ed6915a09b9.metadata.json	NULL	NULL	NULL
2024-01-22 16:51:02.057	hdfs://hadoop-master:9000/metadata/00005-679d6e5b-502d-4f93-a0d2-29d1de0ac01f.metadata.json	4266830935110946003	0	4
Time taken: 13.582 seconds, Fetched 6 row(s)


trino> SELECT * FROM ypoon."iceberg_trino_audit_log_public2$metadata_log_entries";
             timestamp              |                                                               file                                                               | latest_snapshot_id  | latest_schema_id | latest_sequence_number
------------------------------------+----------------------------------------------------------------------------------------------------------------------------------+---------------------+------------------+------------------------
 2024-01-22 16:50:33.208 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00000-acbc6857-46cd-498a-b8e0-95b5a8b661df.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:39.595 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00001-9a9310e3-b54c-4bb1-a9ae-f8a775b51755.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:40.700 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00002-72310e31-4d95-4f60-ba43-9f4e336f1ff9.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:43.645 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00003-ef40bea2-3719-4174-890b-d3aced094ba4.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:44.734 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00004-9698fb96-0ea7-48b4-a499-4ed6915a09b9.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:51:02.057 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00005-679d6e5b-502d-4f93-a0d2-29d1de0ac01f.metadata.json | 4266830935110946003 |                0 |                      4
(6 rows)
```

[the remark this answers is not in the document]
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:94]: The current implementation in Iceberg rely on `snapshot-log` field in metadata file:
https://github.com/apache/iceberg/blob/556b79893c4cf760f031258f0d8ae657e42b4443/core/src/main/java/org/apache/iceberg/MetadataLogEntriesTable.java#L110
https://github.com/apache/iceberg/blob/556b79893c4cf760f031258f0d8ae657e42b4443/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L352
https://github.com/apache/iceberg/blob/556b79893c4cf760f031258f0d8ae657e42b4443/core/src/main/java/org/apache/iceberg/BaseTable.java#L150
`snapshot-log` got reset after after the CREATE OR REPLACE statement.
I can't find a precise definition for `latest_snapshot_id`, `latest_schema_id` and `latest_sequence_number`, so I can't verify is this a correct behavior.
All I know is the result should be the same across different engines since everyone is using `MetadataLogEntriesTable` from Iceberg lib.

[the remark this answers is not in the document]
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:94]: Spark is worse. More records become null. They don't set `snapshot-log` for CREATE and ALTER TABLE.
```
spark-sql> create table test.t1 (c1 integer);
Time taken: 0.27 seconds
spark-sql> alter table test.t1 add columns (c2 string);
Time taken: 0.242 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
Time taken: 1.212 seconds, Fetched 2 row(s)
spark-sql> insert into test.t1 values (1, 'a');
Time taken: 0.423 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	1836642618692808023	1	0
Time taken: 0.14 seconds, Fetched 3 row(s)
spark-sql> delete from test.t1 where c1 = 1;
Time taken: 0.596 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	1836642618692808023	1	0
2024-01-22 18:51:52.6	hdfs://hadoop/tmp/test.db/t1/metadata/00003-806771f6-e7f1-44f5-ac80-350f2c084505.metadata.json	8876099574020403871	1	0
Time taken: 0.121 seconds, Fetched 4 row(s)
spark-sql> CALL local.system.rewrite_data_files('test.t1');
0	0	0
Time taken: 0.173 seconds, Fetched 1 row(s)
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	1836642618692808023	1	0
2024-01-22 18:51:52.6	hdfs://hadoop/tmp/test.db/t1/metadata/00003-806771f6-e7f1-44f5-ac80-350f2c084505.metadata.json	8876099574020403871	1	0
Time taken: 0.106 seconds, Fetched 4 row(s)
spark-sql> create or replace table test.t1 (c3 integer);
Time taken: 0.236 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:52.6	hdfs://hadoop/tmp/test.db/t1/metadata/00003-806771f6-e7f1-44f5-ac80-350f2c084505.metadata.json	NULL	NULL	NULL
2024-01-22 18:52:01.749	hdfs://hadoop/tmp/test.db/t1/metadata/00004-d9c74f49-37c8-4463-8565-7b31660723e2.metadata.json	NULL	NULL	NULL
Time taken: 0.159 seconds, Fetched 5 row(s)
```

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java [hunk 2]
```diff
@@ -301,6 +302,12 @@ public void testSelectSystemTable()
                         .addCopies(GET_TABLE, 1)
                         .build());
 
+        // select from $metadata_log_entries
+        assertMetastoreInvocations("SELECT * FROM \"test_select_snapshots$metadata_log_entries\"",
+                ImmutableMultiset.<MetastoreMethod>builder()
+                        .addCopies(GET_TABLE, 1)
```
Comment (jkylling) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java:308]: nit:
```suggestion
                        .add(GET_TABLE)
```

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 3]
```diff
@@ -7774,6 +7774,14 @@ private List<Long> getTableHistory(String tableName)
                 .collect(toImmutableList());
     }
 
+    private List<Long> getTableMetadataLogEntries(String tableName)
```
Comment (jkylling) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7777]: Is this method used?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static void addRow(PageListBuilder pagesBuilder, StructLike structLike, TimeZoneKey timeZoneKey, Map<String, Integer> columnNameToPositionInSchema)
+    {
+        pagesBuilder.beginRow();
+
+        pagesBuilder.appendTimestampTzMillis(
+                structLike.get(columnNameToPositionInSchema.get(TIMESTAMP_COLUMN_NAME), Long.class) / MICROSECONDS_PER_MILLISECOND,
```
Comment (jkylling) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:128]: Why not let the column have type `TIMESTAMP(6) WITH TIME ZONE`?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    private static final String LATEST_SEQUENCE_NUMBER_COLUMN_NAME = "latest_sequence_number";
+
+    public MetadataLogEntriesTable(SchemaTableName tableName, Table icebergTable)
+    {
+        this.icebergTable = requireNonNull(icebergTable, "icebergTable is null");
+
+        tableMetadata = new ConnectorTableMetadata(requireNonNull(tableName, "tableName is null"),
+                ImmutableList.<ColumnMetadata>builder()
+                        .add(new ColumnMetadata(TIMESTAMP_COLUMN_NAME, TIMESTAMP_TZ_MILLIS))
+                        .add(new ColumnMetadata(FILE_COLUMN_NAME, VARCHAR))
```
Comment (jkylling) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:69]: nit: rename to path or file_path? That seems more in line with the naming in other metadata tables.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java [hunk 6]
```diff
@@ -459,6 +459,12 @@ public void testSelectSystemTable()
                             .addCopies(GET_TABLE, 1)
                             .build());
 
+            // select from $metadata_log_entries
+            assertGlueMetastoreApiInvocations("SELECT * FROM \"test_select_snapshots$metadata_log_entries\"",
```
Comment (jkylling) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:463]: Line 507 needs to be updated?

**Review by jkylling — COMMENTED** [review]
Thanks a lot for adding this!

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java (hunk 4 above)
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:128]: All other system tables are still using `TIMESTAMP(3) WITH TIME ZONE`. We should keep it consistent for better user experience.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java (hunk 2 above)
Comment (oneonestar) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java:308]: Just want to keep it consistent with the codes above and below.

**Review by jkylling — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:94]: @oneonestar what do you plan to do about this bug?
Did you discover the cause of it ? Is it related to the `apache/iceberg` library? If yes, can you pls file an issue about it?
Let's also "document" the erroneous behavior in a test (to know also when it is going to be fixed).

[the remark this answers is not in the document]
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:94]: Asking for a precise definition for `latest_snapshot_id, latest_schema_id, latest_sequence_number` https://github.com/apache/iceberg/issues/9723

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_MILLIS;
+import static io.trino.spi.type.Timestamps.MICROSECONDS_PER_MILLISECOND;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.MetadataTableType.METADATA_LOG_ENTRIES;
+
+public class MetadataLogEntriesTable
+        implements SystemTable
+{
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:53]: I see strong similiarities with `SnapshotsTable`.
Consider creating a common abstract table for code reuse.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 8]
```diff
@@ -6255,18 +6255,23 @@ public void testReadFromVersionedTableWithExpiredHistory()
...
         assertUpdate("INSERT INTO " + tableName + " VALUES (3, 3), (4, 4), (5, 5)", 3);
         List<Long> snapshots = getTableHistory(tableName);
+        List<Long> metadataLogEntries = getTableMetadataLogEntries(tableName);
 
         assertUpdate("DELETE FROM " + tableName + " WHERE c1 < 4", 4);
         List<Long> snapshotsAfterDelete = getTableHistory(tableName);
         assertThat(snapshotsAfterDelete.size()).isGreaterThan(snapshots.size());
         assertThat(snapshotsAfterDelete).containsAll(snapshots);
+
+        List<Long> metadataLogEntriesAfterDelete = getTableMetadataLogEntries(tableName);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:6272]: The changes in this test seem unrelated to the current commit.
Pls add them in a separate commit.

**Review by findinpath — APPROVED** [review]
LGTM % comments

**Comment by findinpath** [comment]
@ebyhr  could you please run this PR with secrets ?
I see changes in `TestIcebergGlueCatalogAccessOperations`.

**Comment by ebyhr** [comment]
/test-with-secrets sha=328415474d00cd4b90a29150616b84bde9f445f2

**Comment by ebyhr** [comment]
/test-with-secrets sha=8aa32a6a8bdb63d586c9fe3179a6ab0980c928e6

**Comment by ebyhr** [comment]
Rebased on master to fix CI failure with secrets.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java (hunk 7 above)
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:53]: @findinpath 
Please take a look at https://github.com/trinodb/trino/commit/11714e731b4e05c66830cb49d6ae3c653dfa7033
Do you like the direction of the refactoring?
Should I add it to this PR or create another PR.

Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:53]: that's what i had in mind.
you could make a preparatory commit to extract the common logic from `SnapshotsTable` which would leave the maintainer to focus on the `metadata_log_entries` changes.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java]: Added a test for this.
This is going to be fixed in Iceberg 1.5.0.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java (hunk 7 above)
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:42]: Added a preparatory commit for refactoring.  [08915d6](https://github.com/trinodb/trino/pull/20410/commits/08915d6b7fa9c6e6412026553d187a0221f98b51)

**Comment by oneonestar** [comment]
Addressed all the comments and rebased to the latest master.
This is ready for review.

File: docs/src/main/sphinx/connector/iceberg.md [hunk 9]
```diff
@@ -825,6 +825,52 @@ The output of the query has the following columns:
...
+:::{list-table} Metadata log entries columns
+:widths: 30, 30, 40
+:header-rows: 1
+
+* - Name
+  - Type
+  - Description
+* - `timestamp`
+  - `TIMESTAMP(3) WITH TIME ZONE`
+  - The time when the metadata was updated.
```
Comment (findinpath) [inline docs/src/main/sphinx/connector/iceberg.md:858]: updated -> created

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 10]
```diff
@@ -7806,6 +7811,14 @@ private List<Long> getTableHistory(String tableName)
                 .collect(toImmutableList());
     }
 
+    private List<Long> getTableMetadataLogEntries(String tableName)
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7814]: The method name is slightly misleading.
The method retrieves the `latest_sequence_number` from the metadata log entries, and not the entire entries.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 10 above)
Comment (oneonestar) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7814]: That's the same as `getTableHistory`. The function is simple enough to read through in a few seconds. And I don't think we want a super long function name. Change it to `getLatestSequenceNumber` is a bit lack of context.
https://github.com/trinodb/trino/blob/1503ecc1182fca9f38780fe5b2e3715d6766ccaf/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java#L7806-L7808

**Comment by oneonestar** [comment]
Rebased to master and resolved conflicts.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Maps.immutableEntry;
+import static com.google.common.collect.Streams.mapWithIndex;
+import static org.apache.iceberg.MetadataTableUtils.createMetadataTableInstance;
+
+public abstract class BaseSystemTable
+        implements SystemTable
+{
+    protected ConnectorTableMetadata tableMetadata;
+    protected Table icebergTable;
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java:47]: Why not initialize these variables in a constructor?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+                        .add(new ColumnMetadata(LATEST_SCHEMA_ID_COLUMN_NAME, INTEGER))
+                        .add(new ColumnMetadata(LATEST_SEQUENCE_NUMBER_COLUMN_NAME, BIGINT))
+                        .build());
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        return new FixedPageSource(buildPages(tableMetadata, session, icebergTable, METADATA_LOG_ENTRIES));
+    }
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java:67]: This method is very similar to SnapshotsTable's one. Why not move to BaseSystemTable?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 13]
```diff
@@ -6308,18 +6308,23 @@ public void testReadFromVersionedTableWithExpiredHistory()
...
+        List<Long> metadataLogEntries = getTableMetadataLogEntries(tableName);
 
         assertUpdate("DELETE FROM " + tableName + " WHERE c1 < 4", 4);
         List<Long> snapshotsAfterDelete = getTableHistory(tableName);
         assertThat(snapshotsAfterDelete.size()).isGreaterThan(snapshots.size());
         assertThat(snapshotsAfterDelete).containsAll(snapshots);
+
+        List<Long> metadataLogEntriesAfterDelete = getTableMetadataLogEntries(tableName);
+        assertThat(metadataLogEntriesAfterDelete.size()).isGreaterThan(metadataLogEntries.size());
+        assertThat(metadataLogEntriesAfterDelete).containsAll(metadataLogEntries);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:6327]: No need to separate assertions:
```java
        assertThat(metadataLogEntriesAfterDelete)
                .hasSizeGreaterThan(metadataLogEntries.size())
                .containsAll(metadataLogEntries);
```

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 10 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7867]: +1 to rename the method, maybe `getLatestSequenceNumbersInMetadataLogEntries`.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java [hunk 14]
```diff
@@ -197,6 +199,83 @@ public void testHistoryTable()
...
+        latestSchemaId.add(2);
+        latestSequenceNumber.add(6L);
+        latestSchemaId.add(2);
+        latestSequenceNumber.add(6L);
+        testMetadataLogEntriesHelper(latestSchemaId, latestSequenceNumber);
+
+        assertUpdate("DROP TABLE IF EXISTS test_schema.test_metadata_log_entries");
+    }
+
+    private void testMetadataLogEntriesHelper(List<Integer> latestSchemaId, List<Long> latestSequenceNumber)
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java:267]: Rename to `assertMetadataLogEntries`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java (hunk 2 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java:308]: No need to use `addCopies` as @jkylling suggested. We can fix other places in follow-up. Same for `TestIcebergGlueCatalogAccessOperations`.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java [hunk 15]
```diff
@@ -197,6 +199,83 @@ public void testHistoryTable()
...
+        assertQuery("SHOW COLUMNS FROM test_schema.\"test_table$metadata_log_entries\"",
+                "VALUES ('timestamp', 'timestamp(3) with time zone', '', '')," +
+                        "('file', 'varchar', '', '')," +
+                        "('latest_snapshot_id', 'bigint', '', '')," +
+                        "('latest_schema_id', 'integer', '', '')," +
+                        "('latest_sequence_number', 'bigint', '', '')");
+
+        List<Integer> latestSchemaId = new ArrayList<>();
+        List<Long> latestSequenceNumber = new ArrayList<>();
+        // Update this test after Iceberg fixed metadata_log_entries after RTAS (https://github.com/apache/iceberg/issues/9723)
```
[the remark this answers is not in the document]
Comment (oneonestar) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java]: https://github.com/apache/iceberg/issues/9723 is correct. It has been fixed and released in Iceberg 1.5.0.
This test needs to be updated after Trino updated to Iceberg 1.5.0.
(see https://github.com/apache/iceberg/pull/9732)

Also, this line is duplicated accidentally. I'll only keep the comments on https://github.com/trinodb/trino/pull/20410/files#diff-2c77e347c613ee81e28953f6b552f710e44bb21e3f75b140770226ac7a1647ceR244-R245.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java (hunk 11 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java]: I don't understand why you initialized only icebergTable here.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        return tableMetadata;
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        return new FixedPageSource(buildPages(tableMetadata, session, icebergTable, getMetadataTableType()));
+    }
+
+    protected abstract MetadataTableType getMetadataTableType();
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java:74]: Move to a constructor.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 17]
```diff
@@ -6320,6 +6320,21 @@ public void testDeleteRetainsTableHistory()
...
+    @Test
+    public void testDeleteRetainsMetadataFile()
+    {
+        String tableName = "test_delete_retains_metadata_file_" + randomNameSuffix();
+        assertUpdate("CREATE TABLE " + tableName + "(c1 INT, c2 INT)");
+        assertUpdate("INSERT INTO " + tableName + " VALUES (1, 1), (2, 2), (3, 3)", 3);
+        assertUpdate("INSERT INTO " + tableName + " VALUES (3, 3), (4, 4), (5, 5)", 3);
+        List<Long> metadataLogEntries = getLatestSequenceNumbersInMetadataLogEntries(tableName);
+
+        List<Long> metadataLogEntriesAfterDelete = getLatestSequenceNumbersInMetadataLogEntries(tableName);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:6334]: DELETE statement looks missing.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java (hunk 11 above)
Comment (oneonestar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/BaseSystemTable.java]: https://github.com/oneonestar/trino/commit/37eaf2fe59de853d2c468dbaa4631821c423c6d0
Tried to make it works.
Base class would be clear but the `super()` call in subclass is a bit complicated.
Do you think this is the way to go?

**Comment by oneonestar** [comment]
OPTIMIZE creates two metadata log entries since https://github.com/trinodb/trino/commit/947f972e09030e0190496b4c98c0c42c8ebc6067

Updated the test accordingly at https://github.com/trinodb/trino/pull/20410/files#diff-2c77e347c613ee81e28953f6b552f710e44bb21e3f75b140770226ac7a1647ceR239-R251

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
@oneonestar Could you rebase on master as https://github.com/trinodb/trino/pull/21034 (Iceberg 1.5.0 bump) landed. Sorry, I missed merging this PR.

**Comment by oneonestar** [comment]
Rebased and updated tests. `latest_snapshot_id` is no longer null after RTAS.

**Comment by oneonestar** [comment]
The error seems unrelated.
